pilot/pkg/xds/debuggen.go

```diff
@@ -0,0 +1,122 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package xds
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+
+	"github.com/golang/protobuf/ptypes/any"
+
+	"istio.io/istio/pilot/pkg/model"
+)
+
+const (
+	// TypeDebug requests debug info from istio, a secured implementation for istio debug interface
+	TypeDebug = "istio.io/debug"
+	debugURL  = "http://localhost:8080/debug/"
+)
+
+var activeNamespaceDebuggers = []string{
+	"config_dump",
+	"ndsz",
+	"syncz",
+	"edsz",
+	"config_distribution",
+}
+
+// DebugGen is a Generator for istio debug info
+type DebugGen struct {
+	Server *DiscoveryServer
+}
+
+func NewDebugGen(s *DiscoveryServer) *DebugGen {
+	return &DebugGen{
+		Server: s,
+	}
+}
+
+// Generate XDS debug responses according to the incoming debug request
+func (dg *DebugGen) Generate(proxy *model.Proxy, push *model.PushContext, w *model.WatchedResource, updates *model.PushRequest) (model.Resources, error) {
+	res := []*any.Any{}
+	var err error
+	var out []byte
+	if w.ResourceNames == nil {
+		res = append(res, &any.Any{
+			TypeUrl: TypeDebug,
+			Value:   []byte("debug type is required"),
+		})
+		return res, nil
+	}
+	if len(w.ResourceNames) != 1 {
+		return res, nil
+	}
+	resourceName := w.ResourceNames[0]
```

Comment:
nit: technically we can support multiple at once. we can add that in the future though